src/grpc/stream_infer_handler.cc

```diff
@@ -619,18 +678,40 @@ ModelStreamInferHandler::StreamInferResponseComplete(
   }
 
   // Update states to signal that response/error is ready to write to stream
-  if (state->is_decoupled_) {
-    std::lock_guard<std::mutex> lock(state->step_mtx_);
-    if (response) {
-      state->response_queue_->MarkNextResponseComplete();
+  {
+    // Need to hold lock because the handler thread processing context
+    // cancellation might have cancelled or marked the state for cancellation.
+    std::lock_guard<std::recursive_mutex> lock(state->step_mtx_);
+
+    if (state->IsGrpcContextCancelled()) {
+      LOG_VERBOSE(1)
+          << "ModelStreamInferHandler::StreamInferResponseComplete, "
+          << state->unique_id_
+          << ", skipping writing response because of transaction was cancelled";
+
+      // If this was the final callback for the state
+      // then cycle through the completion queue so
+      // that state object can be released.
+      if (state->complete_) {
+        state->step_ = Steps::CANCELLED;
+        state->context_->PutTaskBackToQueue(state);
```

Comment:
Why do we only cancel for the final callback? Say a decoupled request has 10 responses that each take a long time, would we not support cancelling after 1 or 2 responses?

Reply:
The other callback invocations would run into segmentation fault or incorrectly point to the state which is now tracking the next request.
It is the callback's responsibility to retrieve all the cancelled invocations are correctly processed.
Additionally, I presume that a response with cancelled state should have FINAL flag set and not issue anymore response callbacks. This would entail that gRPC frontend will cancel the state transaction as soon as it sees the cancelled response with FINAL flag - which can be 1 or 2 response. I didn't want to make this assumption in the gRPC frontend and make it flexible to handle mulitple cases.